File: common/src/java/org/openqa/selenium/Cookie.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium;

import java.util.Date;
import java.text.SimpleDateFormat;

public class Cookie {
    private final String name;
    private final String value;
    private final String path;
    private final String domain;
    private final Date expiry;

    /**
     * Creates a cookie. If the path is left blank or set to null it will be
     * set to "/"
     *
     * @param name name cannot be null or empty string
     * @param value value can be an empty string but not be null
     * @param path The path to use. Will default to "/"
     * @param expiry expiry can be null
     */
    public Cookie(String name, String value, String path, Date expiry) {
      this(name, value, null, path, expiry);
    }

  public Cookie(String name, String value, String domain, String path, Date expiry) {
    this.name = name;
    this.value = value;
    this.path = path == null || "".equals(path) ? "/" : path;
    this.domain = domain;

    if (expiry != null) {
      //igonre the milliseconds because firefox only keeps the seconds
      this.expiry = new Date(expiry.getTime() / 1000 * 1000);
    } else {
      this.expiry = null;
    }

    validate();
  }

  /**
   * Create a cookie for the default path with the given name and value with
   * no expiry set.
   *
   * @param name The cookie's name
   * @param value The cookie's value
   */
    public Cookie(String name, String value) {
        this(name, value, "/", null);
    }

  /**
   * Create a cookie.
   *
   * @param name The cookie's name
   * @param value The cookie's value
   * @param path The path the cookie is for
   */
    public Cookie(String name, String value, String path) {
        this(name, value, path, null);
    }

    public String getName() {
        return name;
    }

    public String getValue() {
        return value;
    }

    public String getDomain() {
        return domain;
    }

    public String getPath() {
        return path;
    }

    public boolean isSecure() {
        return false;
    }

    public Date getExpiry() {
        return expiry;
    }

    protected void validate() {
        if (name == null || "".equals(name) || value == null || path == null)
            throw new IllegalArgumentException("Required attributes are not set or " +
                    "any non-null attribute set to null");

        if (name.indexOf(';') != -1)
            throw new IllegalArgumentException(
                    "Cookie names cannot contain a ';': " + name);
    }

    @Override
    public String toString() {
        return name + "=" + value 
                + (expiry == null ? "" : "; expires=" + new SimpleDateFormat("EEE, dd-MM-yyyy hh:mm:ss z").format(expiry))
                + ("".equals(path) ? "" : "; path=" + path)
                + (domain == null ? "" : "; domain=" + domain);
//                + (isSecure ? ";secure;" : "");
    }

    /**
     *  Two cookies are equal if the name and value match
     */
    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (!(o instanceof Cookie)) {
        return false;
      }

      Cookie cookie = (Cookie) o;

      if (!name.equals(cookie.name)) {
        return false;
      }
      return !(value != null ? !value.equals(cookie.value) : cookie.value != null);
    }

    @Override
    public int hashCode() {
      return name.hashCode();
    }

  public static class Builder {

    private final String name;
    private final String value;
    private String path;
    private String domain;
    private Date expiry;

    public Builder(String name, String value) {
      this.name = name;
      this.value = value;
    }

    public Builder domain(String host) {
      this.domain = host;
      return this;
    }

    public Builder path(String path) {
      this.path = path;
      return this;
    }

    public Builder expiresOn(Date expiry) {
      this.expiry = expiry;
      return this;
    }

    public Cookie build() {
      return new Cookie(name, value, domain, path, expiry);
    }
  }
}

File: common/src/csharp/webdriver-common/Cookie.cs
/* Copyright notice and license
Copyright 2007-2010 WebDriver committers
Copyright 2007-2010 Google Inc.
Portions copyright 2007 ThoughtWorks, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace OpenQA.Selenium
{
    /// <summary>
    /// Represents a cookie in the browser.
    /// </summary>
    public class Cookie
    {
        private string cookieName;
        private string cookieValue;
        private string cookiePath;
        private string cookieDomain;
        private DateTime? cookieExpiry;

        /// <summary>
        /// Initializes a new instance of the <see cref="Cookie"/> class with a specific name, 
        /// value, domain, path and expiration date.
        /// </summary>
        /// <param name="name">The name of the cookie.</param>
        /// <param name="value">The value of the cookie.</param>
        /// <param name="domain">The domain of the cookie.</param>
        /// <param name="path">The path of the cookie.</param>
        /// <param name="expiry">The expiration date of the cookie.</param>
        /// <exception cref="ArgumentException">If the name is <see langword="null"/> or an empty string,
        /// or if it contains a semi-colon.</exception>
        /// <exception cref="ArgumentNullException">If the value is <see langword="null"/>.</exception>
        public Cookie(string name, string value, string domain, string path, DateTime? expiry)
        {
            if (string.IsNullOrEmpty(name))
            {
                throw new ArgumentException("Cookie name cannot be null or empty string", "name");
            }

            if (value == null)
            {
                throw new ArgumentNullException("value", "Cookie value cannot be null");
            }

            if (name.IndexOf(';') != -1)
            {
                throw new ArgumentException("Cookie names cannot contain a ';': " + name, "name");
            }

            this.cookieName = name;
            this.cookieValue = value;
            if (!string.IsNullOrEmpty(path))
            {
                this.cookiePath = path;
            }
            else
            {
                this.cookiePath = "/";
            }

            this.cookieDomain = domain;
            if (expiry != null)
            {
                this.cookieExpiry = expiry;
            }
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Cookie"/> class with a specific name, 
        /// value, path and expiration date.
        /// </summary>
        /// <param name="name">The name of the cookie.</param>
        /// <param name="value">The value of the cookie.</param>
        /// <param name="path">The path of the cookie.</param>
        /// <param name="expiry">The expiration date of the cookie.</param>
        /// <exception cref="ArgumentException">If the name is <see langword="null"/> or an empty string,
        /// or if it contains a semi-colon.</exception>
        /// <exception cref="ArgumentNullException">If the value is <see langword="null"/>.</exception>
        public Cookie(string name, string value, string path, DateTime? expiry)
            : this(name, value, null, path, expiry)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Cookie"/> class with a specific name, 
        /// value, and path.
        /// </summary>
        /// <param name="name">The name of the cookie.</param>
        /// <param name="value">The value of the cookie.</param>
        /// <param name="path">The path of the cookie.</param>
        /// <exception cref="ArgumentException">If the name is <see langword="null"/> or an empty string,
        /// or if it contains a semi-colon.</exception>
        /// <exception cref="ArgumentNullException">If the value is <see langword="null"/>.</exception>
        public Cookie(string name, string value, string path)
            : this(name, value, path, null)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Cookie"/> class with a specific name and value.
        /// </summary>
        /// <param name="name">The name of the cookie.</param>
        /// <param name="value">The value of the cookie.</param>
        /// <exception cref="ArgumentException">If the name is <see langword="null"/> or an empty string,
        /// or if it contains a semi-colon.</exception>
        /// <exception cref="ArgumentNullException">If the value is <see langword="null"/>.</exception>
        public Cookie(string name, string value)
            : this(name, value, "/", null)
        {
        }

        /// <summary>
        /// Gets the name of the cookie.
        /// </summary>
        public string Name
        {
            get { return cookieName; }
        }

        /// <summary>
        /// Gets the value of the cookie.
        /// </summary>
        public string Value
        {
            get { return cookieValue; }
        }

        /// <summary>
        /// Gets the domain of the cookie.
        /// </summary>
        public string Domain
        {
            get { return cookieDomain; }
        }

        /// <summary>
        /// Gets the path of the cookie.
        /// </summary>
        public virtual string Path
        {
            get { return cookiePath; }
        }

        /// <summary>
        /// Gets a value indicating whether the cookie is secure.
        /// </summary>
        public virtual bool Secure
        {
            get { return false; }
        }

        /// <summary>
        /// Gets the expiration date of the cookie.
        /// </summary>
        public DateTime? Expiry
        {
            get { return cookieExpiry; }
        }

        /// <summary>
        /// Creates and returns a string representation of the cookie. 
        /// </summary>
        /// <returns>A string representation of the cookie.</returns>
        public override string ToString()
        {
            return cookieName + "=" + cookieValue
                + (cookieExpiry == null ? string.Empty : "; expires=" + cookieExpiry.Value.ToUniversalTime().ToString("ddd MM/dd/yyyy hh:mm:ss UTC", CultureInfo.InvariantCulture))
                    + (string.IsNullOrEmpty(cookiePath) ? string.Empty : "; path=" + cookiePath)
                    + (string.IsNullOrEmpty(cookieDomain) ? string.Empty : "; domain=" + cookieDomain);
            ////                + (isSecure ? ";secure;" : "");
        }

        /// <summary>
        /// Determines whether the specified <see cref="System.Object">Object</see> is equal 
        /// to the current <see cref="System.Object">Object</see>.
        /// </summary>
        /// <param name="obj">The <see cref="System.Object">Object</see> to compare with the 
        /// current <see cref="System.Object">Object</see>.</param>
        /// <returns><see langword="true"/> if the specified <see cref="System.Object">Object</see>
        /// is equal to the current <see cref="System.Object">Object</see>; otherwise,
        /// <see langword="false"/>.</returns>
        public override bool Equals(object obj)
        {
            // Two cookies are equal if the name and value match
            Cookie cookie = obj as Cookie;

            if (this == obj)
            {
                return true;
            }

            if (cookie == null)
            {
                return false;
            }

            if (!cookieName.Equals(cookie.cookieName))
            {
                return false;
            }

            return !(cookieValue != null ? !cookieValue.Equals(cookie.cookieValue) : cookie.Value != null);
        }

        /// <summary>
        /// Serves as a hash function for a particular type.
        /// </summary>
        /// <returns>A hash code for the current <see cref="System.Object">Object</see>.</returns>
        public override int GetHashCode()
        {
            return cookieName.GetHashCode();
        }
    }
}
